extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/StaticTenantResolver.java

```diff
@@ -358,4 +382,52 @@ private static boolean isNotAnyIssuer(String issuer) {
         }
     }
 
+    private static final class HeaderBasedTenantResolver implements TenantResolver {
+
+        private final Map<String, String> headerNameToTenantId;
+
+        private HeaderBasedTenantResolver(Map<String, String> headerNameToTenantId) {
+            this.headerNameToTenantId = Map.copyOf(headerNameToTenantId);
+        }
+
+        @Override
+        public String resolve(RoutingContext context) {
+            for (String headerName : context.request().headers().names()) {
+                String tenantId = headerNameToTenantId.get(headerName);
+                if (tenantId != null) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debugf("Resolved the '%s' OIDC tenant based on the matching header '%s'", tenantId, headerName);
+                    }
+                    return tenantId;
+                }
+            }
+            return null;
+        }
+
+        private static TenantResolver of(TenantConfigBean tenantConfigBean) {
+            var tenantsWitEnabledHeaderResolution = tenantConfigBean.getStaticTenantsConfig().values().stream()
+                    .map(TenantConfigContext::getOidcTenantConfig)
+                    .filter(Objects::nonNull)
+                    .filter(c -> c.token().header().isPresent())
+                    .filter(c -> !AUTHORIZATION.toString().equalsIgnoreCase(c.token().header().get()))
+                    .toList();
+            if (tenantsWitEnabledHeaderResolution.isEmpty()) {
+                return null;
+            }
+            var headerToTenant = new HashMap<String, String>();
+            for (OidcTenantConfig tc : tenantsWitEnabledHeaderResolution) {
+                String headerName = tc.token().header().get();
+                String tenantId = tc.tenantId().get();
+                String previousTenantId = headerToTenant.put(headerName, tenantId);
+                if (previousTenantId != null) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debugf(
+                                "Cannot resolve OIDC tenant '%s' with the header '%s' as the tenant '%s' is resolved with the same header",
```

Comment:
@michalvavrik Can you please clarify this message ?

Reply:
So, my thoughts were that:

- if there is a duplicate custom HTTP header like `Custom-Authorization`, we should disable that header-based resolver as it will not be useful. As long as there is no duplicate (other than the `Authorization` name), we are fine

As for the issuer-based being higher in priority over the header-based, I don't see a problem with it and I can see your reasoning. I'll do it this afternoon.

Thanks for the input.